### ISSUE TITLE
Quake 2 Brightness Resets on Start

### ISSUE THREAD

**Issue by Quindorrian** [yquake2/yquake2#134]
So, this is a separate issue I am having where the brightness of Yamagi Quake 2 resets to default as soon as the Id Software logo starts the screen pound animation.  At that point, the brightness slider stays where I had it last.  It's almost like the gamma is reset each time, so wherever the slider is becomes the new gamma starting point.  I can then move the slider a tick and it responds where it left off, which is really weird.  Is there anything I can do to fix this?  I am running the latest version of OS X El Capitan.

**Comment by Yamagi** [comment]
This is likely another SDL2 issue on OS X. Quake II doesn't reset the gamma ramp during cinematics playback, but it releases the mouse. So maybe SDL2 is thinking that the application has lost focus.. I would try a build of SDL2 HEAD, the last release is missing the one or another change...

**Comment by Jarvik7** [comment]
Just tested it here. It seems to have nothing to do with cinematics.
To reproduce:
1) Set brightness to maximum in Q2 menu
2) Quit Q2
3) Restart Q2
Results:
1) Attract demo plays at max brightness for about 1 second, then switches to default brightness.
2) Quitting and restarting again will produce the same result (in other words, the saved brightness setting is still max)
3) Dropping the brightness slider down one notch will return to high brightness (max-1)
4) Starting a new game does not change the brightness

System: Fully patched El Capitan running latest commit with shareware pak files (no cinematics!).

**Comment by DanielGibson** [comment]
can you add the following line in `src/backends/sdl/refresh.c` around Line 316 in `UpdateHardwareGamma()`:
`printf("## UpdateHardwareGamma(): %f\n", gamma);`

Now when you start the game from the terminal, each time the game changes the gamma value, there should be a message like "## UpdateHardwareGamma(): 1.0" in the terminal.
Could you check how often you get that message (and with what values) when you start the game?

**Comment by Jarvik7** [comment]
Called only once, before it resets to default brightness
## UpdateHardwareGamma(): 2.000000

**Comment by DanielGibson** [comment]
Default gamma is 1.0, so 2.0 should be pretty bright and match your settings in the menu.

Could you also, in the same file around Line 340, in `CreateSDLWindow()`, add the following line:
`printf("## Creating window in %d x %d\n", vid.width, vid.height);`

Our only remaining theory is that the window is recreated after setting the gamma, so the old gamma settings are lost (though this really shouldn't happen while a demo is being played..).

**Comment by Jarvik7** [comment]
The game starts properly with the correct gamma (2.0), but drops to default after a second of gameplay on screen, so it shouldn't be related to window creation.

**Comment by DanielGibson** [comment]
if the window is recreated after setting the gamma, we probably would have to set gamma for the new window, which I think we don't (we only set gamma if the CVar is marked as changed).  
So if that's not the case, the problem is not our fault but either a bug in SDL or a problem in OSX (maybe some other process resetting screen gamma or something).

**Comment by Jarvik7** [comment]
Bug does not occur when running the game windowed.

**Comment by DanielGibson** [comment]
So CreateSDLWindow() is only called once anyway?

It's good to know it only happens in fullscreen.

**Comment by Quindorrian** [comment]
Thanks guys.  So, you guys able to post the line of code to fix this for full screen mode?  Really prefer full screen to windowed mode.  Thanks again for the help.

**Comment by Jarvik7** [comment]
There is no fix yet. I only gave an idea for a workaround hack.
For now you can either play windowed, or re-set the gamma whenever it gets dark.

**Comment by DanielGibson** [comment]
https://bugzilla.libsdl.org/show_bug.cgi?id=2555 looks like this is an OSX bug:
> It appears that the OS resets the display gamma to the default around 2 seconds after we do a display mode change. I verified that it's not SDL erroneously resetting the gamma.
>
> Also, simply calling Cocoa_SetWindowGammaRamp(_this, window, window->gamma); at the end of Cocoa_SetWindowFullscreen doesn't fix the problem.

**Comment by DanielGibson** [comment]
An OSX user/developer told me that if your display is configured correctly setting gamma shouldn't be necessary anyway and your game/SDL setting gamma (even to 1.0) changes the system gamma settings, which might look crappy then.
Maybe we should just disable the gamma setting on OSX.

**Comment by Quindorrian** [comment]
My OS X is set to default for the iMAC preset and the colors, brightness, and contract are perfect.  In contrast, Quake II's default brightness has never been sufficient on any platform.  So, it requires adjustment by nature.  There are literally dozens of articles online about "my Quake II is too dark".  I would not consider disabling the gamma setting on OS X.  So, if I'm understanding correctly, this is a bug in the engine you guys used to create this Quake 2?

**Comment by DanielGibson** [comment]
The actual bug is in OSX, for resetting HW gamma two seconds after resolution change.

Apart from that, we didn't create the engine, id Software did almost 20 years ago.
People complaining about Quake2 being too dark with whatever port they're using on whatever platform doesn't mean it has to be the case with our port on OSX.

As a dirty hack for testing, could you try in `src/backends/sdl/refresh.c` line 316 (see https://github.com/yquake2/yquake2/blob/master/src/backends/sdl/refresh.c#L316 ) too add "return;", then save the file, compile quake2 again and test with the new executable?
Then vid_gamma shouldn't do anything at all (and this includes **not** changing OSX's preconfigured gamma) - could you check if it's really too dark then?

**Comment by DanielGibson** [comment]
This should work now, using old software gamma code.

The main difference to HW gamma (apart from not being broken on OSX) is that changes to the brightness bar/`vid_gamma` aren't visible immediately, but only after applying them/`vid_restart`

**Comment by Quindorrian** [comment]
Yep, it works now.  Thanks a bunch!  The behavior is slightly different than the original game, but it works.  You have to exit out of the game and reopen the game for changes to take effect.  In the original, it would take effect when you hit apply.  I know the Yamagi Quake 2 implementation was a step further in that the brightness would change in real-time.  Thanks again!
